pkgs/development/interpreters/wasmtime/default.nix

```
@@ -21,11 +21,20 @@ rustPlatform.buildRustPackage rec {
 
   buildInputs = lib.optional stdenv.isDarwin Security;
 
-  # SIMD tests are only executed on platforms that support all
-  # required processor features (e.g. SSE3, SSSE3 and SSE4.1 on x86_64):
-  # https://github.com/bytecodealliance/wasmtime/blob/v9.0.0/cranelift/codegen/src/isa/x64/mod.rs#L220
-  doCheck = with stdenv.buildPlatform; (isx86_64 -> sse3Support && ssse3Support && sse4_1Support);
-  cargoTestFlags = ["--package" "wasmtime-runtime"];
+  # rustfmt is brought into scope to fix the following
+  #   warning: cranelift-codegen@0.108.0:
+  #   Failed to run `rustfmt` on ISLE-generated code: Os
+  #   { code: 2, kind: NotFound, message: "No such file or directory" }
+  nativeBuildInputs = [ rustfmt ];
+
```

As an alternative, have you tried to do instead:

```suggestion

CARGO_TARGET_WASM32_UNKNOWN_UNKNOWN_LINKER = "lld";

```

---

No, let me give that a shot and report back. Thanks for the suggestion!

---

When I build https://github.com/basvandijk/nixpkgs/commit/731e23be27db152d015b51973b325b657efc7aba I get the following error when building the tests on aarch64-darwin:
```
$ nix-build -A wasmtime
...
buildPhase completed in 1 minutes 41 seconds
Running phase: checkPhase
Executing cargoCheckHook
++ cargo test -j 10 --profile release --target aarch64-apple-darwin --frozen -- --test-threads=10
...
      Finished release [optimized] target(s) in 14.77s
  running: env -u CARGO_ENCODED_RUSTFLAGS CARGO_TARGET_DIR="/private/tmp/nix-build-wasmtime-21.0.0.drv-3/source/target/aarch64-apple-darwin/release/build/test-programs-artifacts-2bb492d5129f5bc1/out" RUSTFLAGS="" "cargo" "build" "--release" "--package=wasi-preview1-component-adapter" "--target=wasm32-unknown-unknown" "--no-default-features" "--features=command"
     Compiling wasi-preview1-component-adapter v21.0.0 (/private/tmp/nix-build-wasmtime-21.0.0.drv-3/source/crates/wasi-preview1-component-adapter)
      Finished release [optimized] target(s) in 2.91s
  running: env -u CARGO_ENCODED_RUSTFLAGS CARGO_TARGET_DIR="/private/tmp/nix-build-wasmtime-21.0.0.drv-3/source/target/aarch64-apple-darwin/release/build/test-programs-artifacts-2bb492d5129f5bc1/out" RUSTFLAGS="" "cargo" "build" "--release" "--package=wasi-preview1-component-adapter" "--target=wasm32-unknown-unknown" "--no-default-features" "--features=proxy"
     Compiling wasi-preview1-component-adapter v21.0.0 (/private/tmp/nix-build-wasmtime-21.0.0.drv-3/source/crates/wasi-preview1-component-adapter)
      Finished release [optimized] target(s) in 1.62s
  running: env -u CARGO_ENCODED_RUSTFLAGS CARGO_PROFILE_DEV_DEBUG="1" CARGO_TARGET_DIR="/private/tmp/nix-build-wasmtime-21.0.0.drv-3/source/target/aarch64-apple-darwin/release/build/test-programs-artifacts-2bb492d5129f5bc1/out" RUSTFLAGS="" "cargo" "build" "--target=wasm32-wasi" "--package=test-programs"
...
 error[E0463]: can't find crate for `core`
    |
    = note: the `wasm32-wasi` target may not be installed
    = help: consider downloading the target with `rustup target add wasm32-wasi`

  error[E0463]: can't find crate for `compiler_builtins`

  For more information about this error, try `rustc --explain E0463`.
  error: could not compile `tinyvec_macros` (lib) due to 2 previous errors
  warning: build failed, waiting for other jobs to finish...
  error[E0463]: can't find crate for `core`
    --> /private/tmp/nix-build-wasmtime-21.0.0.drv-3/wasmtime-21.0.0-vendor.tar.gz/typenum/src/bit.rs:13:5
     |
  13 | use core::ops::{BitAnd, BitOr, BitXor, Not};
     |     ^^^^ can't find crate
     |
     = note: the `wasm32-wasi` target may not be installed
     = help: consider downloading the target with `rustup target add wasm32-wasi`

  error[E0463]: can't find crate for `core`
    --> /private/tmp/nix-build-wasmtime-21.0.0.drv-3/wasmtime-21.0.0-vendor.tar.gz/typenum/src/int.rs:37:5
     |
  37 | use core::ops::{Add, Div, Mul, Neg, Rem, Sub};
     |     ^^^^ can't find crate
     |
     = note: the `wasm32-wasi` target may not be installed
     = help: consider downloading the target with `rustup target add wasm32-wasi`
...
```
and many more compile errors after that.

---

This probably means our rustplatform doesn't support the wasm32-wasi target out of the box.

---

Oh, I see, thank for checking. I find it strange that it works under x86_64 though. Something to look into at some point.

---

I see the old code had:
```
cargoTestFlags = ["--package" "wasmtime-runtime"];
```
which was failing since `wasmtime-runtime` doesn't exist anymore.

But maybe I have to limit the tests in a similar way.

---

Building the tests keep tripping over the problematic `wasi-preview1-component-adapter` package. I also tried excluding it via:
```
cargoTestFlags = [ "--workspace" "--exclude" "wasi-preview1-component-adapter" ];
```
but that package is a dependency of something else so it keeps getting pulled in.

I suggest we merge this and try to fix the tests in another PR.

BTW important point: the tests don't actually run for x86_64-linux on ofborg. See: https://logs.ofborg.org/?key=nixos/nixpkgs.314409&attempt_id=a96d2a2f-b6cd-464e-9cbd-ad80e35ad70e. Presumably because those machines don't satisfy `sse3Support && ssse3Support && sse4_1Support`.

---

Yes, this is in purpose. They were failing because they were missing this CPU features due to being virtualized. So adding these checks made them green.

I agree on fixing this, I will take care of cleaning up; it has been a while of me just bumping versions without double checking that everything was sound.

---

Note: the wasmtime-runtime was removed in 72004aad099d235e7e2fc2ae604eb404faff100d (taggedin 21.0.0 for the first time.) I will follow up on this on other PR.